Proper use of `'readable'` event.

mcollina opened the pull request:
Fixes https://github.com/mqttjs/MQTT.js/pull/858

mcollina commented:
@redboltz can you please confirm that this works as expected, and if not, please add a unit test or example with code from just this repo.

redboltz commented:
I tested the PR.  But I experienced the same problem on mqtt-level-store.
It seems that level-db's stream behavior is something different from `Store` on MQTT.js. But I couldn't find what is the difference. So I can't create the test case **from just this repo**.

I can show you the sequence of my test.

As I mentioned, `readable_fail_data_ok_case` branch on my fork is just added the test for demonstrates the problem. 

```
git clone -b readable_fail_data_ok_case https://github.com/redboltz/mqtt-level-store
cd mqtt-level-store
npm install
cp path_to_fix-use-of-readable/lib/client.js node_modules/mqtt/lib/client.js
npm run test
```

Then you can get

```
> mqtt-level-store@2.0.0 test /home/kondo/tmp/mqtt-level-store
> standard && mocha test.js



  mqtt level store
    ✓ should put and stream in-flight packets
    ✓ should support destroying the stream
    ✓ should add and del in-flight packets
    ✓ should replace a packet when doing put with the same messageId
    ✓ should return the original packet on del
    ✓ should get a packet with the same messageId

  mqtt level store manager
    incoming
      ✓ should put and stream in-flight packets
      ✓ should support destroying the stream
      ✓ should add and del in-flight packets
      ✓ should replace a packet when doing put with the same messageId
      ✓ should return the original packet on del
      ✓ should get a packet with the same messageId
    outgoing
      ✓ should put and stream in-flight packets
      ✓ should support destroying the stream
      ✓ should add and del in-flight packets
      ✓ should replace a packet when doing put with the same messageId
      ✓ should return the original packet on del
      ✓ should get a packet with the same messageId

  mqtt.connect flow
Packet {
  cmd: 'publish',
  retain: false,
  qos: 1,
  dup: false,
  length: 16,
  topic: 'hello1',
  payload: <Buffer 77 6f 72 6c 64 31>,
  messageId: 41893 }
Packet {
  cmd: 'publish',
  retain: false,
  qos: 1,
  dup: false,
  length: 16,
  topic: 'hello2',
  payload: <Buffer 77 6f 72 6c 64 32>,
  messageId: 41894 }
Packet {
  cmd: 'publish',
  retain: false,
  qos: 1,
  dup: false,
  length: 16,
  topic: 'hello2',
  payload: <Buffer 77 6f 72 6c 64 32>,
  messageId: 41894 }
    ✓ should resend messages (1060ms)


  19 passing (1s)
```

See the end of the result. The third publish is the same as second one.

redboltz commented:
I noticed something about the problem.

In 

```cpp
    outStore.once('readable', function () {
```

function, 

```cpp
          that._sendPacket(packet)
```

is called.  In `_sendPacket()` function, 'storeAndSend()`, and client.outgoingStore.put()` is called.

This call dispatches level-db's `put()`. This means level-db's `put()` is called during level-db's stream iteration. I'm not sure but it might be a reason of the behavior.

I've tested 'data' approach. In this test case, publish is 6 times called not 3times. It's bad. So I introduced my under developing [mqtt-level-store](https://github.com/redboltz/mqtt-level-store/blob/add_order_preserve_multi_index/mqtt-level-store.js#L67) to `packet.cmd` check on put (replacing case). Replace is needed only 'publish' to 'pubrel' case. This check can avoid the same packet put during iteration.

It requires the following test update. But it is reasonable.

```cpp
  it('should replace a packet when doing put with the same messageId', function (done) {
    var packet1 = {
      cmd: 'publish', // added
      topic: 'hello',
      payload: 'world',
      qos: 2,
      messageId: 42
    }
    var packet2 = {
      cmd: 'pubrel', // added
      qos: 2,
      messageId: 42
    }

    store.put(packet1, function () {
      store.put(packet2, function () {
        store
          .createStream()
          .on('data', function (data) {
            data.should.eql(packet2)
            done()
          })
      })
    })
  })
```

I know this is not directly related this PR. But I guess this information might be helpful to analyze the problem.

mcollina commented:
I've updated this branch, it should fix your problem.

> So I introduced my under developing mqtt-level-store to packet.cmd check on put (replacing case). Replace is needed only 'publish' to 'pubrel' case.

I think that would be needed.

redboltz commented:
@mcollina , thank you very much!! My problem is fixed.
After the PR is merged, I will sent a PR for mqtt-level-store to preserve publish order.

redboltz commented:
When I run test on mqtt-level-store, replace test is failed. (mqtt is correctly updated by npm)
I noticed that the PR **added** the following test case.

```javascript
  it('should replace a packet when doing put with the same messageId', function (done) {
    var packet1 = {
      cmd: 'publish', // added
      topic: 'hello',
      payload: 'world',
      qos: 2,
      messageId: 42
    }
    var packet2 = {
      cmd: 'pubrel', // added
      qos: 2,
      messageId: 42
    }

    store.put(packet1, function () {
      store.put(packet2, function () {
        store
          .createStream()
          .on('data', function (data) {
            data.should.eql(packet2)
            done()
          })
      })
    })
  })
```

But it is existing test case. I only added cmd (just commented // added).
Now, there are two the same name tests.

https://github.com/mqttjs/MQTT.js/blob/master/test/abstract_store.js#L72

https://github.com/mqttjs/MQTT.js/blob/master/test/abstract_store.js#L134

And the first one doesn't have cmd.

redboltz commented:
When I checked the updated PR, I only copied client.js and abstract_store.js is kept as local modification. So I didn't notice earlier.
